### PR TITLE
HPCC-13823 Use NonReentrantSpinLock where possible

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@mckellyln another incremental change to review.  Saves another few % allocating a row.  Please check carefully as ever - I can't see anyway the lock would be requested recursively.

File: roxie/roxiemem/roxiemem.cpp [hunk 1]
```diff
@@ -2764,6 +2764,7 @@ void ChunkedHeaplet::verifySpaceList()
 
 char * ChunkedHeaplet::allocateChunk()
 {
+    //The spin lock for the heap this chunk belongs to must be held when this function is called
```
Comment (mckellyln) [inline roxie/roxiemem/roxiemem.cpp]: Should we, in DEBUG or unittest mode only, add a test to see if the lock for this heap is already set ?

**Comment by mckellyln** [comment]
I doubt it but is there a need for a lock for ChunkedHeaplet::allocate() ?

**Comment by mckellyln** [comment]
Other than my questions looks great for merge.

**Comment by ghalliday** [comment]
allocateChunk() is called from one of two places:

* inlineDoAllocate.  That always holds the lock when calling the function (called in 3 places)

* moveRow, called from compactRow().  state.protectHead must have been called which grabs the lock.

It is possible if heap compaction starts being used in anger that this should become a critical section.

**Comment by mckellyln** [comment]
yes, all good -  @richardkchapman this is good to merge.

**Pull request merged**